GalenAnswers.md

```diff
@@ -0,0 +1,90 @@
+        Involuntary context switches: 1
+        Swaps: 0
+        File system inputs: 0
+        File system outputs: 8
+        Socket messages sent: 0
+        Socket messages received: 0
+        Signals delivered: 0
+        Page size (bytes): 4096
+        Exit status: 0
+```
+This output includes RAM needed (from which one can tell how many cores to request) and files read/written.
+
+3. To alter the previous command so that it writes execution time and RAM useage to a file, one can use the following command instead of using the "-verbose" command:
+```
+/usr/bin/time -f "Time: %E, max RAM: %M" -o timetest.txt echo "This is me testing the time command" >> wctest
+```
+This command prints output to the file "timetest.txt" in the format "Time: x:xx.xx max RAM: xxxx"
+For the command above, it worked out as "Time: 0:00.00, max RAM: 1600"
+
+4. The units of elapsed time are h:mm:ss, and for RAM they are kbytes.
+
+5. One can use the "test" command to see whether files exist using conditional execution or an if...fi statement:
+```
+touch whale
+[ -e whale ] && echo "There she blows" || echo "Argh matey"
+```
+This should output "There she blows" since we just created the file "whale". The "-s" option of the test command will check that a file exists and isn't empty. If we change the "-e" option to "-s", the output will be "Argh matey" since we just created the (empty) file.
+
+6. Here is an example of how an if...fi statement could be used to prevent a command from outputting and overwriting a file which already exists:
+```
+if [ ! -s outfilename.txt ]
+then
+command infilename.txt > outfilename.txt
+else
+echo "stopped: outfilename.txt already exists"
+fi
+```
+
+## Trickier questions
+
+7. "usr/bin/time -v/f/o" will not work on OS X, at least according to [this stack overflow post](https://stackoverflow.com/questions/32515381/mac-os-x-usr-bin-time-verbose-flag) because OS X uses BSD time which doesn't include the RAM measurement. According to an answer on the same post, it can be accomplished by installing GNU time manually. I don't have a computer with OS X, so I haven't been able to test whether my answers are correct.
+
+8. We can ensure that our process won't try to use more RAM than is available and crash the node by ensuring that the node has the necessary amount of free RAM before the job is loaded on:
+```
+#$ -l mem_free=24GB
+```
+In addition, we have to limit the amount of RAM being used by each job. Since each core = 8 Gb, 3 [core ram equivalents](https://hpc.oit.uci.edu/BioClusterGE.pdf) are needed per job (24/8 = 3). So, we add the line:
+```
+-pe openmp 3
+```
+This line will ensure that it is impossible for so many jobs to be loaded onto the node that the node would crash. At most, 21 of these jobs could be loaded onto a single node, using up a total of 504 Gb of RAM at maximum.
+
+Sample script:
+```
+#!/bin/bash
+#$ -N myjobname
+#$ -q bio,free64
+#$ -m beas
+#$ -M galentm@uci.edu   
+#$ -pe openmp 3
+
```

Comment:
Here you should still add the -l mem_free=24GB option. also the # of cores is correct. but you also want to make sure you do the ram/core division for each cluster of nodes. the caveat is that in bio, there are nodes with 256GB and 512GB of ram. for the abio128 nodes there are 128 cores and it has a different ratio. One thing that would have been nice though is that ram per core would be enforced. Unfortunately though it is not, and could be the case for some of your jobs randomly crashing even though you were a good hpc samaritan.